server/map_tree_updater.go

```diff
@@ -110,44 +109,36 @@ func (t *mapTreeUpdater) update(ctx context.Context, tx storage.MapTreeTX, nodes
 }
 
 type txAccessor struct {
-	// hashes is a cache of node hashes that Get returns directly instead of
-	// calling GetMerkleNodes. Used only if preload is true.
-	hashes  map[tree.NodeID2][]byte
-	preload bool
-
-	tx  storage.MapTreeTX
-	rev int64
+	mtu *mapTreeUpdater
+	// read is a cache of tree tiles that Get returns directly instead of calling
+	// GetTiles. It is initialized by the first Get call if not specified.
+	read *smt.TileSet
+	tx   storage.MapTreeTX
+	rev  int64
 }
 
-func (t txAccessor) Get(ctx context.Context, ids []tree.NodeID2) (map[tree.NodeID2][]byte, error) {
+// TODO(pavelkalinnikov): Get should take a list of tile IDs.
+func (t *txAccessor) Get(ctx context.Context, ids []tree.NodeID2) (map[tree.NodeID2][]byte, error) {
 	// TODO(pavelkalinnikov): Factor out preload into another accessor.
-	if t.preload {
-		return t.hashes, nil
-	}
-
-	// TODO(pavelkalinnikov): Pass NodeID2 directly to storage.
-	convIDs := make([]tree.NodeID, 0, len(ids))
-	for _, id := range ids {
-		convIDs = append(convIDs, tree.NewNodeIDFromID2(id))
-	}
-	nodes, err := t.tx.GetMerkleNodes(ctx, t.rev-1, convIDs)
-	if err != nil {
-		return nil, err
-	}
-	res := make(map[tree.NodeID2][]byte, len(nodes))
-	for _, node := range nodes {
-		res[node.NodeID.ToNodeID2()] = node.Hash
+	if t.read == nil {
+		var err error
+		if t.read, err = t.mtu.load(ctx, t.tx, ids, t.rev-1); err != nil {
+			return nil, err
+		}
 	}
-	return res, nil
+	return t.read.Hashes(), nil
 }
 
-func (t txAccessor) Set(ctx context.Context, nodes []smt.Node) error {
-	stNodes := make([]tree.Node, 0, len(nodes))
+func (t *txAccessor) Set(ctx context.Context, nodes []smt.Node) error {
```

Comment:
Is set for leaf nodes or interior nodes? It would be nice to have that documented.

Reply:
Maybe it's the leaf nodes of interior tiles?

Reply:
Added method comment.